### DISCUSSION TITLE
Anyone tired of the breaking changes happened among config files between each update?

### DISCUSSION THREAD

**Discussion by ShikiSuen** [nushell/nushell#9882] [General]
Everytime I update outdated ports in MacPorts, the nushell plagues with issues among configuration files.
Today it even cannot recognize bash, zsh, and even xargs.

![image](https://github.com/nushell/nushell/assets/3164826/0fac3d5c-cbc7-46e7-81a8-c6e29b64bcaa)

Is there a method to lock the version of nushell in MacPorts to one stable release?

**Comment by fdncred** [comment]
Breaking changes should be expected in pre-1.0 software. There is no stable release yet.

`executable was not found` errors are typically because your nushell path wasn't setup, or the env.nu is broken and that part wasn't sourced. With the latest changes we removed `let-env`. So, you'll need to replace those instances with `$env.`. 

Example:

```nushell
let-env FOO = bar
```
becomes
```nushell
$env.FOO = bar
```

**Comment by amtoine** [comment]
@ShikiSuen 
you can use the following script to perform the migration automatically :wink: 
```nushell
# fix Nushell files after upgrading to 0.83.x
#
# # Examples
#     fix the config files
#     > fix
#
#     fix a single file
#     > fix /path/to/file.nu
#
#     fix all the `.nu` files in a directory recursively
#     > fix /path/to/directory/
export def fix [
    path?: path  # the path to a file or a directory with files to fix
]: nothing -> nothing {
    use std log

    def _fix [file: path]: nothing -> nothing {
        log debug $"fixing ($file)"
        open --raw $file | str replace --all --string "$env." "$env." | save --force $file
    }

    let path = $path | default $nu.default-config-dir

    if not ($path | path exists) {
        let span = metadata $path | get span
        error make {
            msg: $"(ansi red_bold)path_not_found(ansi reset)"
            label: {
                text: "no such file or directory"
                start: $span.start
                end: $span.end
            }
        }
    }

    match ($path | path type) {
        "file" => {
            _fix $path
            log info $"($path) has been fixed."
        },
        "dir" => {
            print $path
            ls ($path | path join "**" "*.nu") | each {|it| _fix $it.name}
            log info $"all files in ($path) have been fixed."
        },
        _ => { error make --unspanned {msg: $"(ansi red_bold)unknow internal issue(ansi reset)"} }
    }

    null
}
```

**Comment by panicbit** [comment]
My personal strategy is putting all my config and env changes into separate files and then including them into the config via `source ~/.config/nushell/my-config.nu` and `source ~/.config/nushell/my-env.nu`.
Whenever there's a breaking change I let nu regenerate the config files and then append the source instructions again.

**Comment by siph** [comment]
This may or may not be relevant for you can install `nushell` through `nix` and lock the version that way. You can even lock individual scripts if you're willing to run them with `nix`.

**Reply by Jasha10** [reply]
I am using a [nix-shell shebang](https://nixos.wiki/wiki/Nix-shell_shebang) to invoke a pinned version of nushell to run my nushell script
Here is the script:
```nushell
#! /usr/bin/env nix-shell
#! nix-shell -i nu -p nushell -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/4d513ab5f170d66afa3387bdd718d41aa936ee9f.tar.gz

print $"Here is the current nushell version information: ((version).version)"
```
Here's what happens when I run this script:
```
$ chmod +x ./demo-script.nu 
$ ./demo-script.nu
Here is the current nushell version information: 0.86.0
```
The commit hash [`4d513ab5f170d66afa3387bdd718d41aa936ee9f`](https://github.com/NixOS/nixpkgs/commit/4d513ab5f170d66afa3387bdd718d41aa936ee9f) refers to a recent commit in the nixpkgs repository, effectively pinning nushell to version 0.86.

To pass flags to the nu invocation (e.g. `nu --stdin`), you can modify the second line of the script as follows:
```text
#! nix-shell -i "nu --stdin" -p nushell -I ...
```

**Comment by BatmanAoD** [comment]
I really, really, _really_ want to like nushell enough to use it as my daily driver. This is perhaps the longest I've ever used it as such, actually; it's been my default shell since early September. But this kind of thing is way more frustrating than any level of convenience Nushell brings me compared to, say, Bash. (And I hate Bash!)

I've been trying to use [carapace](https://github.com/rsteube/carapace-bin) for completion, and they added support for `git ls-files` (at my request, even!), so I upgraded Carapace. A recent version of Nu apparently deprecated Carapace's config, so the newer version of Carapace [fixed that issue](https://github.com/rsteube/carapace-bin/issues/1914). But the un-deprecated way wasn't in my version of Nu (0.85, I think), so I had to update Nu as well. And now I get a warning about  `config.cd.abbreviations`, which I'm pretty sure comes from the default config file generated by my previous version of Nu.

The really bizarre thing to me is that `0.85` is _one version behind_ the new version, 0.86. So this really is an "every release" kind of problem. Yes, I understand that pre-1.0 software doesn't have stability guarantees, but... that's _fast_ breakage.

I also realize that each individual fix is probably quite small. In the time it took me to write this up, I could have removed the offending config line. But needing to do that once a month or more is...really just not tenable for me, and I wanted to explain that here since the issue already exists.
